File: backend_go/surface_query/utils/perf_timer.go
package utils

import "time"

type PerfTimer struct {
	startTime time.Time
	lapTime   time.Time
}

func NewPerfTimer() *PerfTimer {
	now := time.Now()
	perfTimer := PerfTimer{startTime: now, lapTime: now}
	return &perfTimer
}

func (pt PerfTimer) Elapsed_s() float64 {
	return time.Since(pt.startTime).Seconds()
}

func (pt PerfTimer) Elapsed_ms() int64 {
	return time.Since(pt.startTime).Milliseconds()
}

func (pt *PerfTimer) Lap_s() float64 {
	elapsed := time.Since(pt.lapTime).Seconds()
	pt.lapTime = time.Now()
	return elapsed
}

func (pt *PerfTimer) Lap_ms() int64 {
	elapsed := time.Since(pt.lapTime).Milliseconds()
	pt.lapTime = time.Now()
	return elapsed
}

File: backend_py/libs/core_utils/webviz_pkg/core_utils/perf_timer.py
import time


class PerfTimer:
    def __init__(self) -> None:
        self._start_s = time.perf_counter()
        self._lap_s = self._start_s

    def elapsed_s(self) -> float:
        """Get the time elapsed since the start, in seconds"""
        return time.perf_counter() - self._start_s

    def elapsed_ms(self) -> int:
        """Get the time elapsed since the start, in milliseconds"""
        return int(1000 * self.elapsed_s())

    def lap_s(self, reset_lap_timer: bool = True) -> float:
        """Get elapsed time since last lap, in seconds"""
        time_now = time.perf_counter()
        elapsed = time_now - self._lap_s

        if (reset_lap_timer):
            self._lap_s = time_now

        return elapsed

    def lap_ms(self, reset_lap_timer: bool = True) -> int:
        """Get elapsed time since last lap, in milliseconds"""
        return int(1000 * self.lap_s(reset_lap_timer=reset_lap_timer))
